### PR TITLE
Implement BigQuery Stress Test

### PR DESCRIPTION
**Pull request by Amar3tto**
This pull request introduces stress tests for BigQueryIO, designed to assess the performance under various conditions. The stress tests simulate dynamic load increases and evaluate the behavior of BigQueryIO for different write formats and methods.

**Changes:**

- Added stress tests for BigQueryIO.
- Implemented dynamic load increases over time to simulate varying workloads.
- Introduced configurations for controlling the stress test parameters, such as the number of columns, write method, write format, etc.
- Added support for exporting metrics to _InfluxDB_ or _BigQuery_ based on the configuration parameter.

**Dynamic load increases over time example:**
image (2)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Amar3tto** [comment]
R: @damccorm @Abacn

File: it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * BigQueryIO stress tests. The test is designed to assess the performance of BigQueryIO under
+ * various conditions.
+ */
```
Comment (Abacn) [inline it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java:84]: Would you mind adding a comment about how can we trigger specific test with gradle command line?

**Review by Abacn — COMMENTED** [review]
This is awesome! Gooing to taking a closer look. Here is a few initial comment

Is the test currently writing metrics to influxDB, if so we can setup a grafana dashboard for it for http://metrics.beam.apache.org/ (in a separate PR) even it is currently empty

File: it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+    /**
+     * Determines the destination for exporting metrics. If set to true, metrics will be exported to
+     * InfluxDB and displayed using Grafana. If set to false, metrics will be exported to BigQuery
+     * and displayed with Looker Studio.
+     */
+    @JsonProperty public boolean exportMetricsToInfluxDB = false;
+
+    /** InfluxDB measurement to publish results to. * */
+    @JsonProperty public String influxMeasurement;
```
Comment (Abacn) [inline it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java:559]: (A note for future dashboard) Would the measurement be unique for each stress test / each IO ?

Pasted an example metrics fyi
```
{
  "TotalStreamingDataProcessed": 0.0,
  "BillableShuffleDataProcessed": 0.0,
  "EstimatedCost": 0.06965586137166667,
  "AvgInputThroughputBytesPerSec": 6.712440411666667E7,
  "ElapsedTime": 1811.0,
  "MaxCpuUtilization": 0.7093815630496711,
  "AvgCpuUtilization": 0.6400222870148885,
  "AvgInputThroughputElementsPerSec": 66132.418,
  "TotalPdUsage": 405457.0,
  "TotalGpuTime": 0.0,
  "TotalSsdUsage": 0.0,
  "MaxInputThroughputElementsPerSec": 71111.15,
  "TotalDcuUsage": 0.0,
  "TotalVcpuTime": 3243.0,
  "TotalShuffleDataProcessed": 0.0,
  "EstimatedDataProcessedGB": 101.5,
  "TotalMemoryUsage": 1.3286034E7,
  "MaxInputThroughputBytesPerSec": 7.217781435E7
}
``` 

As can be seen, the field name obtained from getMetrics is not are the same for all pipelines. If we start to publish to influxDB, we need to consider the naming of measurement field to distinguish different test settings.

Even better is to use influxDB tags to distinguish them, however currently it is not supported by IOITMetrics. 

https://docs.influxdata.com/influxdb/v1/concepts/glossary/#measurement

**Review by Abacn — COMMENTED** [review]
Thanks, this looks pretty good. Just had a comment about the ignored test and metrics nomenclature (no actions need for this PR, something to think about later)

File: it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+              Configuration.fromJsonString(
+                  "{\"numColumns\":20,\"rowsPerSecond\":25000,\"minutes\":240,\"numRecords\":720000000,\"valueSizeBytes\":10000,\"pipelineTimeout\":300,\"runner\":\"DataflowRunner\"}",
+                  Configuration.class));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Test
+  @Ignore
```
[the remark this answers is not in the document]
Comment (Amar3tto) [reply it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java]: Removed `@Ignore `

File: it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java (hunk 2 above)
Comment (Amar3tto) [reply it/google-cloud-platform/src/test/java/org/apache/beam/it/gcp/bigquery/BigQueryIOST.java]: Fixed using suffixes, so the measurement should be unique for each stress test case.

It's a good idea to use tags, we'll look into that in the future.

**Review by Abacn — APPROVED** [review]
Thanks!

**Comment by akashorabek** [comment]
[Write job result](https://console.cloud.google.com/dataflow/jobs/us-central1/2024-04-25_07_05_49-11278409782781130961;step=s8;graphView=0?project=apache-beam-testing&pageState=(%22dfTime%22:(%22l%22:%22dfJobMaxTime%22)))

**Pull request merged**